src/bakery/Bakery.cpp

```diff
@@ -188,17 +188,143 @@ void Bakery::greedyWithDijkstra(Van& van) {
+    while (i > 0) {
+        if (table[i][w] - table[i - 1][w - clients[i - 1]->getBreadQuantity()] == values[i - 1]) {
+            // Element i should be allocated
+            v.addClient(clients[i - 1]);
+            removed++;
+            w -= clients[i - 1]->getBreadQuantity();
+        }
+        i--;
+    }
+    return removed;
+}
+
+int Bakery::greedyAllocation(Van &v) {
+    int count = 0;
+    int capacity = v.getTotalBread();
+
+    while (true) {
+        int highestValue = -INF;
+        Client *chosen = NULL;
+
+        /*
+         * Value = 50000 + 10 * numBreads - 100 * Euclidean(start, pos)
+         * Summation(10 * abs(v.clients[i].time - time) - 100 * Euclidean(v.clients[i].pos, pos))
+         */
+        for (Client *client : clients) {
+            if (client->isAllocated()) continue;
+
+            int value = 50000 + 10 * client->getBreadQuantity() - 100 *
+                    startingVertex->getPosition().distance(client->getVertex()->getPosition());
+
+            for (Client *client2 : v.getClients())
+                value += abs((client2->getRealTime() - client->getRealTime()).toMinutes())
+                        - 100 * client2->getVertex()->getPosition().distance(client->getVertex()->getPosition());
+
+            if (value > highestValue && client->getBreadQuantity() <= capacity) {
+                highestValue = value;
+                chosen = client;
+            }
+        }
+        if (chosen == NULL) break;
+        v.addClient(chosen);
+        count++;
+        capacity -= chosen->getBreadQuantity();
+    }
+    return count;
+}
+
+void Bakery::allocateClientsToVans(bool useKnapsack, bool optimize) {
+    sort(vans.begin(), vans.end(), [](const Van& v1, const Van& v2) -> bool {
+        return v1.getTotalBread() > v2.getTotalBread();
+    });
+
+    int clientsAllocated = 0;
+
+    if (useKnapsack) {
+        // Value = breadQuantity / 10 + 1 (the client itself has value)
+        vector<int> values;
+        for (Van &v : vans) {
+            if (clientsAllocated == clients.size()) break;
+
+            for (Client *client : clients)
+                values.push_back(client->getBreadQuantity() + 10);
+
+            clientsAllocated += knapsackAllocation(v, values);
+            values.clear();
+        }
+    } else {
+        for (Van &v : vans) {
+            if (clientsAllocated == clients.size()) break;
+            clientsAllocated += greedyAllocation(v);
+        }
+    }
+
+    if (optimize) {
+        /* Since the last used van probably has space left,
+         * it will try to take some clients from other vans
+         */
+        for (int i = vans.size() - 1; i >= 0; --i)
+            if (!vans[i].getClients().empty()) {
+                for (int j = i - 1; j >= 0; --j) {
+                    Client *client = vans[j].removeFarthestClientInRange(vans[i].getReservedBread());
+                    if (client != NULL) vans[i].addClient(client);
```

Comment:
Maybe we could have a calculateSimulation function that would check whether the change improves the overall time. That way we could add more than 1 client from the vans that are full. Maybe we do this on another pr tho